src/client/encode.rs

```diff
@@ -1,108 +1,106 @@
 
         // Insert Proxy-Connection header when method is CONNECT
-        if req.method() == Method::Connect {
-            let val = "proxy-connection: keep-alive\r\n".to_owned();
-            log::trace!("> {}", &val);
-            buf.write_all(val.as_bytes()).await?;
+        if self.request.method() == Method::Connect {
+            self.request.insert_header("proxy-connection", "keep-alive");
         }
 
         // If the body isn't streaming, we can set the content-length ahead of time. Else we need to
         // send all items in chunks.
-        if let Some(len) = req.len() {
-            let val = format!("content-length: {}\r\n", len);
-            log::trace!("> {}", &val);
-            buf.write_all(val.as_bytes()).await?;
+        if let Some(len) = self.request.len() {
+            self.request.insert_header(CONTENT_LENGTH, len.to_string());
         } else {
-            // write!(&mut buf, "Transfer-Encoding: chunked\r\n")?;
-            panic!("chunked encoding is not implemented yet");
-            // See: https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Transfer-Encoding
-            //      https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Trailer
+            self.request.insert_header(TRANSFER_ENCODING, "chunked");
         }
 
-        for (header, values) in req.iter() {
-            for value in values.iter() {
-                let val = format!("{}: {}\r\n", header, value);
-                log::trace!("> {}", &val);
-                buf.write_all(val.as_bytes()).await?;
+        Ok(())
+    }
+
+    fn compute_head(&mut self) -> io::Result<Cursor<Vec<u8>>> {
+        let mut buf = Vec::with_capacity(100);
+        let url = self.request.url();
+        let method = self.request.method();
+        write!(buf, "{} ", method)?;
+
+        // A client sending a CONNECT request MUST consists of only the host
+        // name and port number of the tunnel destination, separated by a colon.
+        // See: https://tools.ietf.org/html/rfc7231#section-4.3.6
+        if method == Method::Connect {
+            let host = url
+                .host_str()
+                .ok_or_else(|| io::Error::new(io::ErrorKind::InvalidData, "Missing hostname"))?;
+
+            let port = url.port_or_known_default().ok_or_else(|| {
+                io::Error::new(
+                    io::ErrorKind::InvalidData,
+                    "Unexpected scheme with no default port",
+                )
+            })?;
+
+            write!(buf, "{}:{}", host, port)?;
+        } else {
+            write!(buf, "{}", url.path())?;
+            if let Some(query) = url.query() {
+                write!(buf, "?{}", query)?;
             }
         }
 
-        buf.write_all(b"\r\n").await?;
+        write!(buf, " HTTP/1.1\r\n")?;
+
+        self.finalize_headers()?;
+        let mut headers = self.request.iter().collect::<Vec<_>>();
+        headers.sort_unstable_by_key(|(h, _)| if **h == HOST { "0" } else { h.as_str() });
```

Comment:
It seems to me that we'd be better off using some kind of order-stable hashmap rather that doing a sort here, since this sort could be `O(n log(n))`. Not sure if a good one exists though.

Reply:
yeah… this logic also probably belongs in http-types, but as mentioned elsewhere i think that should be a distinct move from this change